### PR TITLE
[Design] Reduce DI Allocations

### LINKED ISSUE #288: Reduce DI Allocations
For 2000 plantext requests DI allocates 16,009 objects:

image

Clocking in at 788,304 bytes:

image

Which would mean if you were performing 1M rps the GC would have to flush 8,004,500 objects and 394 Mbytes per second, 788 Mbytes per second for 2M rps and 2.76GBytes (56M objects) per second for 7M rps (~10GbE saturation)

### PR DESCRIPTION
**Pull request by benaadams**
Resolves #288

### CODE REVIEW

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs [hunk 1]
```diff
@@ -19,12 +19,12 @@ internal class ServiceProvider : IServiceProvider, IDisposable
     {
         private readonly object _sync = new object();
 
-        private readonly Func<Type, Func<ServiceProvider, object>> _createServiceAccessor;
+        private Func<Type, Func<ServiceProvider, object>> _createServiceAccessor;
```
Comment (davidfowl) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: why this change?

Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: In error, has no state

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs [hunk 2]
```diff
@@ -8,20 +8,37 @@ namespace Microsoft.Framework.DependencyInjection.ServiceLookup
     internal class ServiceScope : IServiceScope
     {
         private readonly ServiceProvider _scopedProvider;
+        private readonly ServiceScopeFactory _parentFactory;
 
         public ServiceScope(ServiceProvider scopedProvider)
         {
             _scopedProvider = scopedProvider;
         }
+        internal ServiceScope(ServiceProvider scopedProvider, ServiceScopeFactory parentFactory)
```
Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Why internal for this? will the public constructor ever be used with these changes?

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScopeFactory.cs [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Concurrent;
+
 namespace Microsoft.Framework.DependencyInjection.ServiceLookup
 {
     internal class ServiceScopeFactory : IServiceScopeFactory
     {
+        private static int _maxPooled = 32 * Environment.ProcessorCount;
```
Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScopeFactory.cs]: `_capacity`?

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs [hunk 4]
```diff
@@ -145,15 +145,12 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
 
         public void Dispose()
         {
-            var disposables = Interlocked.Exchange(ref _disposables, null);
```
Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: This looks like it was done to avoid a double-dispose if `Dispose()` is called concurrently. With the current code it won't double-dispose any services but could call `Clear()` twice - is this a concern?

I suspect we're not too concerned about this scenario in practice. If there's any code you have operating on an SP while it's being disposed, you already have a bug.

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs [hunk 5]
```diff
@@ -8,20 +8,37 @@ namespace Microsoft.Framework.DependencyInjection.ServiceLookup
...
+        }
 
         public IServiceProvider ServiceProvider
         {
             get { return _scopedProvider; }
         }
 
+        internal void Reset()
+        {
+            // Double wash
```
Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Can you expand on this? what does this mean?

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs [hunk 6]
```diff
@@ -145,15 +145,12 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
...
+            IDisposable disposable;
+            while (_disposables.TryTake(out disposable))
             {
-                foreach (var disposable in disposables)
-                {
-                    disposable.Dispose();
-                }
+                disposable.Dispose();
             }
+            _resolvedServices.Clear();
```
Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: This needs a unit test - something where we can prove that that the following works correctly:

- Create a scoped SP
- Resolve a scoped service (instance 1)
- Dispose scoped SP
- Resolve a scoped service (instance 2)
- Assert that 1 and 2 aren't the same instance

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs (hunk 2 above)
Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Suppose not, missed it was an internal class

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScopeFactory.cs (hunk 3 above)
Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScopeFactory.cs]: Changed

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs (hunk 5 above)
Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Dispose again before reuse incase any extra disposables were added - overkill probably, but am concerned about information leakage between objects.

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs (hunk 4 above)
Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: It would just empty any extras; or do nothing if there were none. I have a little concern as its a concurrentbag which suggests items can be added in parallel; or out of order and with delays.

Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: In fact next comment is about the forced double dispose.

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs (hunk 5 above)
Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Phrased another way, original dispose is to drop any references so the can be GC'ed as the object is now kept; second Reset dispose is to make it safe for reuse.

I could swap ```_disposables``` to a backing object to make any adds error; then restore it in the Reset instead?

Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Made this change

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs [hunk 7]
```diff
@@ -19,9 +21,19 @@ public IServiceProvider ServiceProvider
             get { return _scopedProvider; }
         }
 
+        internal void Reset()
```
Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Why internal here? It would also be helpful to capture your explanation of this in comments.

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs [hunk 8]
```diff
@@ -143,16 +144,28 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
             }
         }
 
+        internal void Reset()
+        {
+            if (_offlineDisposables == null)
+            {
+                throw new InvalidOperationException("ServiceProvider reused before being disposed");
```
Comment (rynowak) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: Do we have a resx in this project? Programmatic names like `ServiceProvider` need to be formatted in with parameters.

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs (hunk 7 above)
Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: Not sure what to write... Its a method that is explicitly bonded to the pooling of ServiceScopeFactory so if the class were made public rather than internal for whatever reason the exposed api would be for it to function as a standalone class (which is why I left public constructor in); whereas the "internals" only work as a class pair.

Not hugely fussed on visibility though, can make public :)

File: src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs (hunk 8 above)
Comment (benaadams) [inline src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs]: Done?

**Comment by benaadams** [comment]
*Note:* still need to implement [unit test](https://github.com/aspnet/DependencyInjection/pull/289#discussion_r39432328) proposed by @rynowak, lost in *outdated diff* messages

File: src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs (hunk 7 above)
Comment (davidfowl) [inline src/Microsoft.Framework.DependencyInjection/ServiceLookup/ServiceScope.cs]: I would leave it internal

**Comment by benaadams** [comment]
See https://github.com/aspnet/HttpAbstractions/issues/402#issuecomment-140336990 for details on server run improvements (61.5k rps -> 1.49M rps)

**Comment by rynowak** [comment]
:watch: - just waiting on a unit test for the scenario we discussed

**Comment by benaadams** [comment]
Have the test, just trying to think of the right name XD

**Comment by halter73** [comment]
:+1:
